File: LibsodiumSignatureDetachedString/LibsodiumSignatureDetachedString.java
package Libsodium_signature_detached_string;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

public class LibsodiumSignatureDetachedString {
    // uses TweetNacl https://github.com/InstantWebP2P/tweetnacl-java
    public static void main(String[] args) {
        System.out.println("Libsodium detached signature string");

        String dataToSign = "The quick brown fox jumps over the lazy dog";
        System.out.println("plaintext: " + dataToSign);

        // usually we would load the private and public key from a file or keystore
        // here we use hardcoded keys for demonstration - don't do this in real programs

        System.out.println("* * * sign the plaintext with the ED25519 private key * * *");
        byte [] privateKey = loadEd25519PrivateKey();
        String signatureBase64 = libsodiumSignDetachedToBase64(privateKey, dataToSign);

        System.out.println("signature (Base64): " + signatureBase64);
        System.out.println("signature expected: x41mufah/9VO347W+nPXu5FYeSJOI894YClbbTiX0pwRrvfAPMymxEvyMMsDFMI0R0sulCnuCRSgN0WOKnZBDg==");
        System.out.println("\n* * * verify the signature against the plaintext with the ED25519 public key * * *");
        byte[] publicKey = loadEd25519PublicKey();
        System.out.println("signature (Base64): " + signatureBase64);
        boolean signatureVerified = libsodiumVerifyDetachedFromBase64(publicKey, dataToSign, signatureBase64);
        System.out.println("signature (Base64) verified: " + signatureVerified);
    }

    private static String libsodiumSignDetachedToBase64(byte[] privateKey, String dataToSign) {
        TweetNaclFast.Signature sig = new TweetNaclFast.Signature(new byte[32], privateKey);
        return base64Encoding(sig.detached(dataToSign.getBytes(StandardCharsets.UTF_8)));
    };

    private static boolean libsodiumVerifyDetachedFromBase64(byte[] publicKey, String dataToSign, String signatureBase64) {
        TweetNaclFast.Signature sig = new TweetNaclFast.Signature(publicKey, new byte[64]);
        return sig.detached_verify(dataToSign.getBytes(StandardCharsets.UTF_8), base64Decoding(signatureBase64));
    }

    private static byte[] loadEd25519PrivateKey() {
        // this is a sample key - don't worry !
        return base64Decoding("Gjho7dILimbXJY8RvmQZEBczDzCYbZonGt/e5QQ3Vro2sjTp95EMxOrgEaWiprPduR/KJFBRA+RlDDpkvmhVfw==");
    }

    private static byte[] loadEd25519PublicKey() {
        // this is a sample key - don't worry !
        return base64Decoding("NrI06feRDMTq4BGloqaz3bkfyiRQUQPkZQw6ZL5oVX8=");
    }

    private static String base64Encoding(byte[] input) {return Base64.getEncoder().encodeToString(input);
    }
    private static byte[] base64Decoding(String input) {return Base64.getDecoder().decode(input);
    }
}

File: LibsodiumSignatureDetachedString/LibsodiumSignatureDetachedString.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter_sodium/flutter_sodium.dart';

main() {
/* add in pubspec.yaml:
dependencies:
  flutter_sodium: ^0.2.0
 */

  printC('Libsodium detached signature string');
  final dataToSign = 'The quick brown fox jumps over the lazy dog';
  printC('plaintext: ' + dataToSign);

  // usually we would load the private and public key from a file or keystore
  // here we use hardcoded keys for demonstration - don't do this in real programs

  // signature
  printC('\n* * * sign the plaintext with the ED25519 private key * * *');
  var privateKey = loadEd25519PrivateKey();
  var signatureBase64 = libsodiumSignDetachedToBase64(privateKey, dataToSign);
  printC('signature Base64: ' + signatureBase64);

  // verification
  printC('\n* * * verify the signature against the plaintext with the ED25519 public key * * *');
  var publicKey = loadEd25519PublicKey();
  printC('signature Base64: ' + signatureBase64);
  var signatureVerified = libsodiumVerifyDetachedFromBase64(publicKey, dataToSign, signatureBase64);
  printC('signature (Base64) verified: ' + signatureVerified.toString());

}

String libsodiumSignDetachedToBase64(Uint8List privateKey, String dataToSign) {
  var signature = CryptoSign.signStringDetached(dataToSign, privateKey);
  return base64Encoding(signature);
}

bool libsodiumVerifyDetachedFromBase64(Uint8List publicKey, String dataToSign, String signatureBase64) {
  var signature = base64Decoding(signatureBase64);
  return CryptoSign.verifyString(signature, dataToSign, publicKey);
}

Uint8List loadEd25519PrivateKey() {
  return base64Decoding("Gjho7dILimbXJY8RvmQZEBczDzCYbZonGt/e5QQ3Vro2sjTp95EMxOrgEaWiprPduR/KJFBRA+RlDDpkvmhVfw==");
}

Uint8List loadEd25519PublicKey() {
  return base64Decoding("NrI06feRDMTq4BGloqaz3bkfyiRQUQPkZQw6ZL5oVX8=");
}

Uint8List createUint8ListFromString(String s) {
  var ret = new Uint8List(s.length);
  for (var i = 0; i < s.length; i++) {
    ret[i] = s.codeUnitAt(i);
  }
  return ret;
}

String base64Encoding(Uint8List input) {
  return base64.encode(input);
}

Uint8List base64Decoding(String input) {
  return base64.decode(input);
}

void printC(String newString) {
  // for compatibility reasons to FlutterEmptyConsole
  print(newString);
}
